docs/tutorials/basic/data.md

```diff
@@ -125,37 +39,59 @@ for batch in data_iter:
 
-## Read CSV
+    [[<NDArray 30x3 @cpu(0)>], [<NDArray 30 @cpu(0)>], 0]
+    [[<NDArray 30x3 @cpu(0)>], [<NDArray 30 @cpu(0)>], 0]
+    [[<NDArray 30x3 @cpu(0)>], [<NDArray 30 @cpu(0)>], 0]
+    [[<NDArray 30x3 @cpu(0)>], [<NDArray 30 @cpu(0)>], 20L]
+
+
+
+## Reading Data from CSV Files
+*MXNet* provides [`CSVIter`](http://mxnet.io/api/python/io.html#mxnet.io.CSVIter) to read from CSV Files and can be used as below:
 
-There is an iterator called to `CSVIter` to read data batches from CSV files. We
-first dump `data` into a csv file, and then load the data
 
 ```python
+#lets save `data` into a csv file first and try reading it back
 np.savetxt('data.csv', data, delimiter=',')
 data_iter = mx.io.CSVIter(data_csv='data.csv', data_shape=(3,), batch_size=30)
 for batch in data_iter:
     print([batch.data, batch.pad])
 ```
 
-Note that we have not given a label file, then `batch.label` is empty here.
+    [[<NDArray 30x3 @cpu(0)>], 0]
+    [[<NDArray 30x3 @cpu(0)>], 0]
+    [[<NDArray 30x3 @cpu(0)>], 0]
+    [[<NDArray 30x3 @cpu(0)>], 20]
 
-## Read images
 
-<!-- TODO(mli) move notebooks here -->
 
-- [Read images](https://github.com/dmlc/mxnet-notebooks/blob/master/python/basic/image_io.ipynb)
-- [Advanced image reading](https://github.com/dmlc/mxnet-notebooks/blob/master/python/basic/advanced_img_io.ipynb)
+## Custom  Iterator
+When the in-built iterators does not suit your application, you can create a custom data iterator.
 
-## Write your own data iterators
+An iterator in _MXNet_ should  
+1. Implement `next()` in ``Python2`` or `__next()__` in ``Python3``,   
+   returning `DataBatch` or raise `StopIteration` exception at the end of the datastream.  
+2. Implement `reset()` method to restart reading from the beginning.   
+3. Have `provide_data` attribute, returning a list of `DataDesc` objects, 
+   described [here](http://mxnet.io/api/python/io.html#mxnet.io.DataBatch).  
+4. Have `provide_label` attribute, returns similar to `provide_label` information about input labels.  
+
+You can either create a iterator from scratch by defining `DataBatch` and the iterator, one such example is show below or reuse existing iterators to create new iterator. For example, in the image caption application, the input example is an image while the label is a sentence. The we can create a new Iterator by:
+- creating image_iter using `ImageRecordIter` which provides multithreaded pre-fetch and augmentation.
+- creating caption_iter using `NDArrayIter` or bucketing iterator provided in the rnn package.
+- `next()` returns the combined result of `image_iter.next()` and `caption_iter.next()`
 
-Sometimes the provided iterators are not enough for some application. There are
-mainly two ways to develop a new iterator. One is creating from scratch, the
-following codes define an iterator that creates a given number of data batches
-through a data generator `data_gen`.
 
 ```python
 import numpy as np
-class SimpleIter:
+
+class SimpleBatch(mx.io.DataBatch):
```

Comment:
@piiswrong, this is very simple example to show that you can have a specialized class of DataBatch and use it in the Iterator, agree it isn't doing much here.